docs/adr/0003-threadsafe-interfaces.md

```diff
@@ -0,0 +1,196 @@
+trying to ship this as an initial fix for Project Nimbus.
+
+### Option D: Use a non-locking handlemap and rely on calling code to behave safely.
+
+Implement a variant of `ConcurrentHandleMap` that does not protect its members with a Mutex, which
+is the main source of constraints on concurrent execution in the current setup. Instead, this handlemap
+would hand out references without any runtime checks, on the assumption that the calling code is
+behaving in a safe manner. Replace all currently uses of `ConcurrentHandleMap` with this new non-locking variant.
+
+* Good, because this is a relatively small change from the current behaviour, which should be fairly
+  quick to ship.
+* Bad, because we lost Rust's compile-time safety guarantees.
+* Bad, because it's hard to communicate to consumers what "behave safely" really means (and we might not
+  even understand it ourselves).
+
+This seems to throw away some of the key safety benefits of using Rust, which makes it a very
+unattractive option.
+
+## Implementation Sketch
+
+### Add a `[Threadsafe]` attribute to UDL `interface` definitions.
+
+We would advise Nimbus SDK to update their UDL definition for `NimbusClient` like so:
+
+```idl
+[Threadsafe]
+interface NimbusClient {
+  // existing method definitions remain unchanged
+}
+```
+
+The [`uniffi_bindgen::interface::Object`](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/interface/mod.rs#L722) struct
+would grow a corresponding boolean `threadsafe` field and corresponding public accessor.
+When [building an instance of this struct from the UDL](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/interface/mod.rs#L786),
+we would inspect the list of attributes to look for one named "Threadsafe" and set the field
+if present. The way that we [handle the `[ByRef]` annotation on method arugments](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/interface/mod.rs#L658)
+will likely serve as a good example to follow.
+
+### Implement a less-locking HandleMap variant
+
+The `ffi_support` crate provides a basic non-locking [`HandleMap<T>`](https://docs.rs/ffi-support/0.4.0/ffi_support/handle_map/struct.HandleMap.html) struct,
+and it implements `ConcurrentHandleMap` as a thin wrapper around a `RwLock<HandleMap<Mutex<T>>>`. We can make our own variant that removes the inner mutex, as a thin wrapper around a`RwLock<HandleMap<T>>`. (The outer `RwLock` is still needed in order to guard mutations of the handlemap itself, and we must be careful to only hold it for short periods).
+
+Bikeshed name: `NonLockingHandleMap`, implemented under the [`uniffi::ffi` module](https://github.com/mozilla/uniffi-rs/tree/main/uniffi/src/ffi). It needs to live in the `uniffi` crate so that it can be used by the generated Rust scaffolding at runtime.
+
```

Comment:
How about 'LessLockingHandleMap' or 'MinimalLockingHandleMap', since there _is_ still a lock in there, and projects that don't require a singleton (i.e. not Nimbus) could still potentially block on locks during object construction/destruction?  If this is too edge-casey to scare all code readers about, it'd be nice to at least document the possibility in the reference.

Reply:
FWIW I will not block (hah!) this on the selection of any particular name; all of the suggestions so far are acceptable to me and I'll be happy to have a verbose name if it conveys an accurate mental model of how the thing behaves. (Which I feel like `ConcurrentHandleMap` doesn't necessarily do - is it "concurrent" because it's safe to use with concurrent access, or because it enables lots of concurrency?)

Reply:
Agreed, the existing name is itself bad - so my 2c would be something like "NonConcurrent..." - ie, if you understand how the term "concurrent" applies to the existing one, you automagically understand what's different about the new one.